osu.Server.Spectator/Hubs/SpectatorHub.cs

```diff
@@ -1,27 +1,78 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Caching.Distributed;
+using osu.Game.Beatmaps;
 using osu.Game.Online.Spectator;
+using osu.Game.Scoring;
+using osu.Game.Scoring.Legacy;
+using osu.Server.Spectator.Database;
 using osu.Server.Spectator.Entities;
 
 namespace osu.Server.Spectator.Hubs
 {
     public class SpectatorHub : StatefulUserHub<ISpectatorClient, SpectatorClientState>, ISpectatorServer
     {
-        public SpectatorHub(IDistributedCache cache, EntityStore<SpectatorClientState> users)
+        public const string REPLAYS_PATH = "replays";
+
+        private readonly IDatabaseFactory databaseFactory;
+
+        public SpectatorHub(IDistributedCache cache, EntityStore<SpectatorClientState> users, IDatabaseFactory databaseFactory)
             : base(cache, users)
         {
+            this.databaseFactory = databaseFactory;
         }
 
         public async Task BeginPlaySession(SpectatorState state)
         {
             using (var usage = await GetOrCreateLocalUserState())
             {
-                usage.Item ??= new SpectatorClientState(Context.ConnectionId, CurrentContextUserId);
-                usage.Item.State = state;
+                var clientState = (usage.Item ??= new SpectatorClientState(Context.ConnectionId, CurrentContextUserId));
+
+                if (clientState.State != null)
+                {
+                    // Previous session never received EndPlaySession call.
+                    // Should probably be handled in some way.
+                }
+
+                clientState.State = state;
+
+                if (state.RulesetID == null) throw new ArgumentNullException(nameof(state.RulesetID));
+                if (state.BeatmapID == null) throw new ArgumentNullException(nameof(state.BeatmapID));
+
+                using (var db = databaseFactory.GetInstance())
+                {
+                    string? beatmapChecksum = await db.GetBeatmapChecksumAsync(state.BeatmapID.Value);
+                    string? username = await db.GetUsernameAsync(CurrentContextUserId);
+
+                    if (string.IsNullOrEmpty(beatmapChecksum))
+                        throw new ArgumentException(nameof(state.BeatmapID));
```

Comment:
Probably would be a good idea to also send the checksum from the client and verify that it matches as well, but I guess it is not a huge priority for now?

Reply:
I did notice this, but I think this is best left for a separate effort.